"""
Reference free AF-based demultiplexing on pooled scRNA-seq
Jon Xu (jun.xu@uq.edu.au)
Lachlan Coin
Aug 2018
"""

import vcf  # https://pyvcf.readthedocs.io/en/latest/INTRO.html
import numpy as np
import pandas as pd
from scipy.stats import binom
from scipy.sparse import csr_matrix
import datetime
import csv


def main():

    # input and output files
    ori_vcf = 'input8.vcf'        # original vcf file used in demuxlet run
    ref_csv = 'ref_sim8.csv'      # reference matrix
    alt_csv = 'alt_sim8.csv'      # alternative matrix
    psc_csv = 'P_s_c.csv'         # P(s|c) generated by sc_split
    dem_grp = 'd'                 # prefix of demuxlet assignments

    ref = pd.read_csv(ref_csv, header=0, index_col=0)
    alt = pd.read_csv(alt_csv, header=0, index_col=0)
    ref_s = csr_matrix(ref.values)
    alt_s = csr_matrix(alt.values)
    all_POS = ref.index

    # build matrix from scSplit assignment
    P_s_c = pd.read_csv(psc_csv, header=0, index_col=0)
    A_s_c = ((P_s_c >= 0.9) * 1).astype('float64')
    num = len(P_s_c.columns)  # number of samples + 1 doublet state
    err = 0.01  # error rate assumption
    # binomial simulation for genotype likelihood P(D|AA,RA,RR) with the alt count vs total count condition and (err, 0.5, 1-err) as allele probability
    rr = pd.DataFrame(binom.pmf(pd.DataFrame(alt_s.dot(A_s_c)), pd.DataFrame((alt_s + ref_s).dot(A_s_c)), err), index=all_POS, columns=range(num)).drop(0,1).apply(np.log10)
    ra = pd.DataFrame(binom.pmf(pd.DataFrame(alt_s.dot(A_s_c)), pd.DataFrame((alt_s + ref_s).dot(A_s_c)), 0.5), index=all_POS, columns=range(num)).drop(0,1).apply(np.log10)
    aa = pd.DataFrame(binom.pmf(pd.DataFrame(alt_s.dot(A_s_c)), pd.DataFrame((alt_s + ref_s).dot(A_s_c)), 1-err), index=all_POS, columns=range(num)).drop(0,1).apply(np.log10)
    rr = 1 / (1 + 10 ** (ra - rr) + 10 ** (aa - rr))
    ra = 1 / (1 + 10 ** (rr - ra) + 10 ** (aa - ra))
    aa = 1 / (1 + 10 ** (rr - aa) + 10 ** (ra - aa))   
    rr = (rr > 0.99) * 1
    ra = (ra > 0.99) * 2
    aa = (aa > 0.99) * 3
    scsplit = rr + ra + aa
    
    # build matrix from demuxlet assignment
    rr[:] = ra[:] = aa[:] = 0    
    for n in range(1, num):
        dem_result = []
        for line in open(dem_grp+str(n), 'r'):
            dem_result.append(line.strip())  # barcodes for each sample
        rr.loc[:, n] = np.log10(binom.pmf(alt.loc[:,dem_result].sum(axis=1), (alt+ref).loc[:,dem_result].sum(axis=1), err))
        ra.loc[:, n] = np.log10(binom.pmf(alt.loc[:,dem_result].sum(axis=1), (alt+ref).loc[:,dem_result].sum(axis=1), 0.5))
        aa.loc[:, n] = np.log10(binom.pmf(alt.loc[:,dem_result].sum(axis=1), (alt+ref).loc[:,dem_result].sum(axis=1), 1-err))
    rr = 1 / (1 + 10 ** (ra - rr) + 10 ** (aa - rr))
    ra = 1 / (1 + 10 ** (rr - ra) + 10 ** (aa - ra))
    aa = 1 / (1 + 10 ** (rr - aa) + 10 ** (ra - aa))   
    rr = (rr > 0.99) * 1
    ra = (ra > 0.99) * 2
    aa = (aa > 0.99) * 3
    demuxlet = rr + ra + aa

    # build matrix from original vcf
    rr[:] = ra[:] = aa[:] = 0    
    for record in vcf.Reader(open(ori_vcf, 'r')):
        for n in range (1, num):
            if record.samples[n-1]['GP'][0] > 0.99:
                rr.loc[record.CHROM + ':' + str(record.POS), n] = 1
            elif record.samples[n-1]['GP'][1] > 0.99:
                rr.loc[record.CHROM + ':' + str(record.POS), n] = 2
            elif record.samples[n-1]['GP'][2] > 0.99:
                rr.loc[record.CHROM + ':' + str(record.POS), n] = 3
    vcf_all = rr + ra + aa

    # output
    scsplit.to_csv('verify_scsplit.csv')
    demuxlet.to_csv('verify_demuxlet.csv')
    vcf_all.to_csv('verify_original.csv')

if __name__ == '__main__':
    main()
